README.md
# Model Optimization

- [Task](#Task)
- [Dataset](#dataset)
  - [TACO](#taco)
  - [Evaluation](#evaluation)
- [Training details](#training-details)
- [Result](#result)
- [Usage](#usage)
  - [Install requirements](#install-requirements)
  - [Train](#train)
  - [Tune](#tune)
  - [Inference](#inference)

---

## Task
최근들어 여러 산업에서 인공지능을 이용해 그동안 해결하지 못 했던 문제들을 풀려는 노력을 하고 있습니다. 대표적인 태스크로는 쓰레기의 재활용 가능 여부을 판단하는 인공지능 쓰레기통이 있습니다. 이를 위해서는 쓰레기 이미지를 분류하는 모델이 탑재되어야 합니다. 하지만 분류만 잘 한다고 해서 사용할 수 있는 것은 아닙니다. 로봇 내부 시스템에 탑재되어 즉각적으로 쓰레기를 분류할 수 있어야만 실제로 사용이 될 수 있습니다.

본 프로젝트(2021.05.24 - 2021.06.18)에서는 분리수거 로봇의 기초 기술인 쓰레기 이미지 분류기를 만들면서 실제로 로봇에 탑재될 만큼 작고 계산량이 적은 모델을 만들었습니다. Base model로는 [MobileNetV2](https://arxiv.org/abs/1801.04381), [EfficientNet](https://arxiv.org/abs/1905.11946) 를 사용했고, Optuna AutoML로 모델 경량화 및 최적화를 진행했습니다.

## Dataset
### TACO
Segmentation / Object detecion task를 풀기 위해 제작된 COCO format의 재활용 쓰레기 데이터인 TACO 데이터셋을 사용합니다. 단순한 Classification 문제로 설정하기 위해 TACO의 Bounding box를 crop한 데이터를 사용했습니다.

![TACO](https://user-images.githubusercontent.com/77161691/125893628-6319352f-b26c-4b65-a7ea-32fff671030e.png)

TACO dataset은 총 9개의 category로 이루어져 있습니다.

```python
category = ["Battery", "Clothing", "Glass", "Metal", "Paper", "Paperpack", "Plastic", "Plasticbag", "Styrofoam"]
```

본 프로젝트에서 사용하는 데이터의 분포는 아래와 같습니다.

![Category](https://user-images.githubusercontent.com/77161691/125894094-e03d1246-78fe-44c4-a9c2-2cfe115b5f8a.png)

- train dataset: 총 32,599장 (label 포함)
- test dataset: 총 8,159장 (label 미포함)

### Evaluation
본 프로젝트에서는 `f1-score`와 `MACs`의 두 가지 척도를 도입했습니다. `f1-score`는 이미지의 추론된 라벨과 실제 라벨이 얼마나 일치하는지를 측정하는 `모델 성능`에 관한 지표입니다. `MACs`는 모델의 총 연산량을 나타내는 `모델 경량화`에 관한 지표입니다.

일정 `f1-score` 이상을 만족한 후에는 로봇에 탑재될 수 있을 만큼 `MACs`를 줄이는 것이 중요합니다. 현실에서의 효용성 평가를 위해 아래와 같이 `total score`로 모델을 평가했습니다.

![Score](https://user-images.githubusercontent.com/77161691/125894785-5b0df7e9-9c91-4fa9-b8c4-735c9e00abd9.png)

## Training Details
| Model | Image Size | Batch Size | Epochs | Initial LR | Train & Optimize Time |
| :--- | ---: | ---: | ---: | ---: | ---: |
| `Efficientnet` | 70 | 64 | 500 | 0.2 | 5d |
| `MobileNetV2` | 224 | 256 | 200 | 0.1 | 2d |

- **P40**을 이용하여 학습하였습니다.

## Result
| Model | Total Score | F1 Score | MACs |
| --- | ---: | ---: | ---: |
| [EfficientNet](https://arxiv.org/abs/1905.11946) | 0.7064 | 0.5275 | 4670834 |
| [MobileNetV2](https://arxiv.org/abs/1801.04381) | 9.9846 | 0.6525 | 135389513 |

## Usage
### Install requirements
Please refer to `environment.yml`

### Train
```
python train.py
```
Please refer to `train.py` for train arguments.

### Tune
```
python tune.py
```
Please refer to `tune.py` for evaluation arguments.

### Inference
```
python inference.py
```
Please refer to `inference.py` for inference arguments.

---
#### Docs
- https://arxiv.org/abs/1905.11946
- https://arxiv.org/abs/1801.04381
- https://optuna.readthedocs.io/en/stable/tutorial/index.html
- https://arxiv.org/abs/1909.13719v2


history.txt
sub1
effinetb0 , epoch : 50
Model saved. Current best test f1: 0.521
Train: [050] Loss: 1.176, Acc: 58.94% F1(macro): 0.43
 Val:       Loss: 1.030, Acc: 64.21% F1(macro): 0.51:
 Val:       Loss: 1.027, Acc: 64.26% F1(macro): 0.52:
Result
score : 3.4050
f1 : 0.5106
macs : 41809542
------------------------------------------------------------------------------------------------------
sub2
effinetb0 , epoch : 50
img_size = 112
Model saved. Current best test f1: 0.438
Train: [050] Loss: 1.302, Acc: 54.68% F1(macro): 0.36
Val:       Loss: 1.153, Acc: 59.76% F1(macro): 0.43
Val:       Loss: 1.152, Acc: 59.93% F1(macro): 0.44
score : 1.7499
f1 : 0.4526
macs : 10397222
------------------------------------------------------------------------------------------------------
sub3
effinetb0 , epoch : 50
img_size = 56
Model saved. Current best test f1: 0.360
Train: [050] Loss: 1.436, Acc: 49.48% F1(macro): 0.29
Val:       Loss: 1.279, Acc: 55.14% F1(macro): 0.35
Val:       Loss: 1.282, Acc: 55.06% F1(macro): 0.36
score : 1.1842
f1 : 0.3655
macs : 2553606
------------------------------------------------------------------------------------------------------
sub4
effinetb0 , epoch : 100
img_size = 56
Model saved. Current best test f1: 0.412
Train: [100] Loss: 1.330, Acc: 53.54% F1(macro): 0.34
Val:       Loss: 1.176, Acc: 59.13% F1(macro): 0.41
score : 1.1842
f1 : 0.4133
macs : 2553606
------------------------------------------------------------------------------------------------------
sub5
myeffnetb0 , epoch :200
img_size = 31
[
    # [repeat, module, args]
       [1, Conv, [32, 3, 2]],
       # repeat(n), MBConv: t, c, s, k
       [1, MBConv, [1, 16, 1, 3]],
       [2, MBConv, [6, 24, 2, 3]],
       [2, MBConv, [6, 40, 2, 5]],
       [3, MBConv, [6, 80, 2, 3]],
       [3, MBConv, [6, 112, 1, 5]],
       [4, MBConv, [6, 192, 2, 5]],
       [1, MBConv, [6, 320, 1, 3]],
       [1, Conv, [128, 1, 1,3]],
       [1, GlobalAvgPool, []],
       [1, Conv, [9, 1, 1, null, 1, null]]
    ]
Model saved. Current best test f1: 0.364
Train: [131] Loss: 1.462, Acc: 48.22% F1(macro): 0.29:
 Val:       Loss: 1.286, Acc: 54.80% F1(macro): 0.36:
------------------------------------------------------------------------------------------------------
sub6
DATA_PATH: "/opt/ml/input/data/"
DATASET: "TACO"
IMG_SIZE: 31
AUG_TRAIN: "randaugment_train"
AUG_TEST: "simple_augment_test"
AUG_TRAIN_PARAMS:
    n_select: 2
    level: null
BATCH_SIZE: 128
EPOCHS: 200
VAL_RATIO: 0.0
INIT_LR: 0.1
FP16: True
idx |   n |     params |          module |            arguments |   in_channel |   out_channel
----------------------------------------------------------------------------------------------
  0 |   1 |        928 |            Conv |           [32, 3, 2] |            3           32
  1 |   1 |      1,176 |          MBConv |        [1, 16, 1, 3] |           32            8
  2 |   2 |      3,300 |          MBConv |        [6, 24, 2, 3] |            8            8
  3 |   2 |      4,836 |          MBConv |        [6, 40, 2, 5] |            8            8
  4 |   3 |      4,950 |          MBConv |        [6, 80, 2, 3] |            8            8
  5 |   3 |      7,254 |          MBConv |       [6, 112, 1, 5] |            8            8
  6 |   4 |      9,672 |          MBConv |       [6, 192, 2, 5] |            8            8
  7 |   1 |      1,650 |          MBConv |       [6, 320, 1, 3] |            8            8
  8 |   1 |      1,280 |            Conv |          [128, 1, 1] |            8          128
  9 |   1 |          0 |   GlobalAvgPool |                   [] |          128          128
 10 |   1 |      2,080 |            Conv | [9, 1, 1, None, 1, None] |          128           16
Model Summary: 352 layers, 37,126 parameters, 37,126 gradients
Model save path: exp/2021-06-07_13-18-21/best.pt
macs: 847494.0
------------------------------------------------------------------------------------------------------
sub8
AUG_TEST: simple_augment_test
AUG_TRAIN: randaugment_train
AUG_TRAIN_PARAMS:
  level: null
  n_select: 2
BATCH_SIZE: 64
DATASET: TACO
DATA_PATH: /opt/ml/input/data/
EPOCHS: 200
FP16: true
IMG_SIZE: 56
INIT_LR: 0.1
VAL_RATIO: 0.0
idx |   n |     params |          module |            arguments |   in_channel |   out_channel
----------------------------------------------------------------------------------------------
  0 |   1 |        928 |            Conv |           [32, 3, 2] |            3           32
  1 |   1 |      1,176 |          MBConv |        [1, 16, 1, 3] |           32            8
  2 |   1 |      1,650 |          MBConv |        [6, 24, 2, 3] |            8            8
  3 |   1 |      1,650 |          MBConv |        [6, 40, 2, 3] |            8            8
  4 |   1 |      1,650 |          MBConv |        [6, 80, 2, 3] |            8            8
  5 |   1 |      1,650 |          MBConv |       [6, 112, 1, 3] |            8            8
  6 |   1 |      1,650 |          MBConv |       [6, 192, 2, 3] |            8            8

tune.py
import argparse
import copy
import optuna
import os
from datetime import datetime
import torch
import torch.nn as nn
import torch.optim as optim
from src.dataloader import create_dataloader
from src.model import Model
from src.utils.torch_utils import model_info
from src.utils.common import read_yaml
from src.utils.macs import calc_macs
from src.trainer import TorchTrainer
from typing import Any, Dict, List, Tuple, Union
from train import train


MODEL_CONFIG = read_yaml(cfg="configs/model/effinetb1.yaml")
DATA_CONFIG = read_yaml(cfg="configs/data/taco.yaml")


def search_hyperparam(trial: optuna.trial.Trial) -> Dict[str, Any]:
    """Search hyperparam from user-specified search space."""
    # epochs = trial.suggest_int("epochs", low=400, high=600, step=100)
    img_size = trial.suggest_int("img_size", low=42, high=98, step=14)
    # n_select = trial.suggest_int("n_select", low=1, high=2, step=1)
    batch_size = trial.suggest_int("batch_size", low=32, high=128, step=32)
    # "EPOCHS": epochs,
    #"n_select": n_select,

    return {
        "IMG_SIZE": img_size,
        "BATCH_SIZE": batch_size
    }


def search_model(trial: optuna.trial.Trial) -> List[Any]:
    """Search model structure from user-specified search space."""
    model = []
    n_stride = 0
    MAX_NUM_STRIDE = 5
    UPPER_STRIDE = 2

    # Module 1
    m1 = trial.suggest_categorical("m1", ["Conv", "DWConv"])
    m1_args = []
    m1_repeat = trial.suggest_int("m1/repeat", 1, 3)
    m1_out_channel = trial.suggest_int("m1/out_channels", low=16, high=64, step=16)
    m1_stride = trial.suggest_int("m1/stride", low=1, high=UPPER_STRIDE)
    if m1_stride == 2:
        n_stride += 1
    m1_activation = trial.suggest_categorical(
        "m1/activation", ["ReLU", "Hardswish"]
        )
    if m1 == "Conv":
        # Conv args: [out_channel, kernel_size, stride, padding, groups, activation]
        m1_args = [m1_out_channel, 3, m1_stride, None, 1, m1_activation]
    elif m1 == "DWConv":
        # DWConv args: [out_channel, kernel_size, stride, padding_size, activation]
        m1_args = [m1_out_channel, 3, m1_stride, None, m1_activation]
    model.append([m1_repeat, m1, m1_args])

    # Module 2
    m2 = trial.suggest_categorical(
        "m2",
        ["Conv",
        "DWConv",
        "MBConv",
        "InvertedResidualv2",
        "InvertedResidualv3",
        "Pass"]
        )
    m2_args = []
    m2_repeat = trial.suggest_int("m2/repeat", 1, 5)
    m2_out_channel = trial.suggest_int("m2/out_channels", low=16, high=128, step=16)
    m2_stride = trial.suggest_int("m2/stride", low=1, high=UPPER_STRIDE)
    # force stride m2
    if n_stride == 0:
        m2_stride = 2
    if m2 == "Conv":
        # Conv args: [out_channel, kernel_size, stride, padding, groups, activation]
        m2_kernel = trial.suggest_int("m2/kernel_size", low=1, high=5, step=2)
        m2_activation = trial.suggest_categorical("m2/activation", ["ReLU", "Hardswish"])
        m2_args = [m2_out_channel, m2_kernel, m2_stride, None, 1, m2_activation]
    elif m2 == "DWConv":
        # DWConv args: [out_channel, kernel_size, stride, padding_size, activation]
        m2_kernel = trial.suggest_int("m2/kernel_size", low=1, high=5, step=2)
        m2_activation = trial.suggest_categorical("m2/activation", ["ReLU", "Hardswish"])
        m2_args = [m2_out_channel, m2_kernel, m2_stride, None, m2_activation]
    elif m2 == "MBConv":
        m2_c = trial.suggest_int("m2/c", low=16, high=320, step=16)
        m2_k = trial.suggest_int("m2/k", low=3, high=5, step=2)
        m2_args = [1, m2_c, m2_stride, m2_k]
    elif m2 == "InvertedResidualv2":
        m2_c = trial.suggest_int("m2/v2_c", low=16, high=32, step=16)
        m2_t = trial.suggest_int("m2/v2_t", low=1, high=4)
        m2_args = [m2_c, m2_t, m2_stride]
    elif m2 == "InvertedResidualv3":
        m2_kernel = trial.suggest_int("m2/kernel_size", low=3, high=5, step=2)
        m2_t = round(trial.suggest_float("m2/v3_t", low=1.0, high=6.0, step=0.1), 1)
        m2_c = trial.suggest_int("m2/v3_c", low=16, high=40, step=8)
        m2_se = trial.suggest_categorical("m2/v3_se", [0, 1])
        m2_hs = trial.suggest_categorical("m2/v3_hs", [0, 1])
        # k t c SE HS s
        m2_args = [m2_kernel, m2_t, m2_c, m2_se, m2_hs, m2_stride]
    if not m2 == "Pass":
        if m2_stride == 2:
            n_stride += 1
            if n_stride>=MAX_NUM_STRIDE:
                UPPER_STRIDE = 1
        model.append([m2_repeat, m2, m2_args])

    # Module 3
    m3 = trial.suggest_categorical(
        "m3",
        ["Conv",
        "DWConv",
        "MBConv",
        "InvertedResidualv2",
        "InvertedResidualv3",
        "Pass"]
        )
    m3_args = []
    m3_repeat = trial.suggest_int("m3/repeat", 1, 5)
    m3_out_channel = trial.suggest_int("m3/out_channels", low=16, high=128, step=16)
    m3_stride = trial.suggest_int("m3/stride", low=1, high=UPPER_STRIDE)
    if m3 == "Conv":
        # Conv args: [out_channel, kernel_size, stride, padding, groups, activation]
        m3_out_channel = trial.suggest_int("m3/out_channels", low=16, high=128, step=16)
        m3_kernel = trial.suggest_int("m3/kernel_size", low=1, high=5, step=2)
        m3_activation = trial.suggest_categorical("m3/activation", ["ReLU", "Hardswish"])
        m3_args = [m3_out_channel, m3_kernel, m3_stride, None, 1, m3_activation]
    elif m3 == "DWConv":
        # DWConv args: [out_channel, kernel_size, stride, padding_size, activation]
        m3_out_channel = trial.suggest_int("m3/out_channels", low=16, high=128, step=16)
        m3_kernel = trial.suggest_int("m3/kernel_size", low=1, high=5, step=2)
        m3_activation = trial.suggest_categorical("m3/activation", ["ReLU", "Hardswish"])
        m3_args = [m3_out_channel, m3_kernel, m3_stride, None, m3_activation]
    elif m3 == "MBConv":
        m3_c = trial.suggest_int("m3/c", low=16, high=320, step=16)
        m3_k = trial.suggest_int("m3/k", low=3, high=5, step=2)
        m3_args = [6, m3_c, m3_stride, m3_k]
    elif m3 == "InvertedResidualv2":
        m3_c = trial.suggest_int("m3/v2_c", low=8, high=32, step=8)
        m3_t = trial.suggest_int("m3/v2_t", low=1, high=8)
        m3_args = [m3_c, m3_t, m3_stride]
    elif m3 == "InvertedResidualv3":
        m3_kernel = trial.suggest_int("m3/kernel_size", low=3, high=5, step=2)
        m3_t = round(trial.suggest_float("m3/v3_t", low=1.0, high=6.0, step=0.1), 1)
        m3_c = trial.suggest_int("m3/v3_c", low=8, high=40, step=8)
        m3_se = trial.suggest_categorical("m3/v3_se", [0, 1])
        m3_hs = trial.suggest_categorical("m3/v3_hs", [0, 1])
        m3_args = [m3_kernel, m3_t, m3_c, m3_se, m3_hs, m3_stride]
    if not m3 == "Pass":
        if m3_stride == 2:
            n_stride += 1
            if n_stride>=MAX_NUM_STRIDE:
                UPPER_STRIDE = 1
        model.append([m3_repeat, m3, m3_args])

    # Module 4
    m4 = trial.suggest_categorical(
        "m4",
        ["Conv",
        "DWConv",
        "MBConv",
        "InvertedResidualv2",
        "InvertedResidualv3",
        "Pass"]
        )
    m4_args = []
    m4_repeat = trial.suggest_int("m4/repeat", 1, 5)
    m4_out_channel = trial.suggest_int("m4/out_channels", low=16, high=128, step=16)
    m4_stride = trial.suggest_int("m4/stride", low=1, high=UPPER_STRIDE)
    # force stride m4
    if n_stride == 1:
        m4_stride = 2
    if m4 == "Conv":
        # Conv args: [out_channel, kernel_size, stride, padding, groups, activation]
        m4_out_channel = trial.suggest_int("m4/out_channels", low=16, high=256, step=16)
        m4_kernel = trial.suggest_int("m4/kernel_size", low=1, high=5, step=2)
        m4_activation = trial.suggest_categorical("m4/activation", ["ReLU", "Hardswish"])
        m4_args = [m4_out_channel, m4_kernel, m4_stride, None, 1, m4_activation]
    elif m4 == "DWConv":
        # DWConv args: [out_channel, kernel_size, stride, padding_size, activation]
        m4_out_channel = trial.suggest_int("m4/out_channels", low=16, high=256, step=16)
        m4_kernel = trial.suggest_int("m4/kernel_size", low=1, high=5, step=2)
        m4_activation = trial.suggest_categorical("m4/activation", ["ReLU", "Hardswish"])
        m4_args = [m4_out_channel, m4_kernel, m4_stride, None, m4_activation]
    elif m4 == "MBConv":
        m4_c = trial.suggest_int("m4/c", low=16, high=320, step=16)
        m4_k = trial.suggest_int("m4/k", low=3, high=5, step=2)
        m4_args = [6, m4_c, m4_stride, m4_k]
    elif m4 == "InvertedResidualv2":
        m4_c = trial.suggest_int("m4/v2_c", low=8, high=64, step=8)
        m4_t = trial.suggest_int("m4/v2_t", low=1, high=8)
        m4_args = [m4_c, m4_t, m4_stride]
    elif m4 == "InvertedResidualv3":
        m4_kernel = trial.suggest_int("m4/kernel_size", low=3, high=5, step=2)
        m4_t = round(trial.suggest_float("m4/v3_t", low=1.0, high=6.0, step=0.1), 1)
        m4_c = trial.suggest_int("m4/v3_c", low=8, high=80, step=8)
        m4_se = trial.suggest_categorical("m4/v3_se", [0, 1])
        m4_hs = trial.suggest_categorical("m4/v3_hs", [0, 1])
        m4_args = [m4_kernel, m4_t, m4_c, m4_se, m4_hs, m4_stride]
    if not m4 == "Pass":
        if m4_stride == 2:
            n_stride += 1
            if n_stride>=MAX_NUM_STRIDE:
                UPPER_STRIDE = 1
        model.append([m4_repeat, m4, m4_args])

    # Module 5
    m5 = trial.suggest_categorical(
        "m5",
        ["Conv",
        "DWConv",
        "MBConv",
        "InvertedResidualv2",
        "InvertedResidualv3",
        "Pass"]
        )
    m5_args = []
    m5_repeat = trial.suggest_int("m5/repeat", 1, 5)
    m5_out_channel = trial.suggest_int("m5/out_channels", low=16, high=128, step=16)
    m5_stride = 1
    if m5 == "Conv":
        # Conv args: [out_channel, kernel_size, stride, padding, groups, activation]
        m5_out_channel = trial.suggest_int("m5/out_channels", low=16, high=256, step=16)
        m5_kernel = trial.suggest_int("m5/kernel_size", low=1, high=5, step=2)
        m5_activation = trial.suggest_categorical("m5/activation", ["ReLU", "Hardswish"])
        m5_stride = trial.suggest_int("m5/stride", low=1, high=UPPER_STRIDE)
        m5_args = [m5_out_channel, m5_kernel, m5_stride, None, 1, m5_activation]
    elif m5 == "DWConv":
        # DWConv args: [out_channel, kernel_size, stride, padding_size, activation]
        m5_out_channel = trial.suggest_int("m5/out_channels", low=16, high=256, step=16)
        m5_kernel = trial.suggest_int("m5/kernel_size", low=1, high=5, step=2)
        m5_activation = trial.suggest_categorical("m5/activation", ["ReLU", "Hardswish"])
        m5_stride = trial.suggest_int("m5/stride", low=1, high=UPPER_STRIDE)
        m5_args = [m5_out_channel, m5_kernel, m5_stride, None, m5_activation]
    elif m5 == "MBConv":
        m5_c = trial.suggest_int("m5/c", low=16, high=320, step=16)
        m5_k = trial.suggest_int("m5/k", low=3, high=5, step=2)
        m5_args = [6, m5_c, m5_stride, m5_k]
    elif m5 == "InvertedResidualv2":
        m5_c = trial.suggest_int("m5/v2_c", low=16, high=128, step=16)
        m5_t = trial.suggest_int("m5/v2_t", low=1, high=8)
        m5_stride = trial.suggest_int("m5/stride", low=1, high=UPPER_STRIDE)
        m5_args = [m5_c, m5_t, m5_stride]
    elif m5 == "InvertedResidualv3":
        m5_kernel = trial.suggest_int("m5/kernel_size", low=3, high=5, step=2)
        m5_t = round(trial.suggest_float("m5/v3_t", low=1.0, high=6.0, step=0.1), 1)
        m5_c = trial.suggest_int("m5/v3_c", low=16, high=80, step=16)
        m5_se = trial.suggest_categorical("m5/v3_se", [0, 1])
        m5_hs = trial.suggest_categorical("m5/v3_hs", [0, 1])
        m5_stride = trial.suggest_int("m5/stride", low=1, high=UPPER_STRIDE)
        m5_args = [m5_kernel, m5_t, m5_c, m5_se, m5_hs, m5_stride]
    if not m5 == "Pass":
        if m5_stride == 2:
            n_stride += 1
            if n_stride>=MAX_NUM_STRIDE:
                UPPER_STRIDE = 1
        model.append([m5_repeat, m5, m5_args])

    # Module 6
    m6 = trial.suggest_categorical(
        "m6",
        ["Conv",
        "DWConv",
        "MBConv",
        "InvertedResidualv2",
        "InvertedResidualv3",
        "Pass"]
        )
    m6_args = []
    m6_repeat = trial.suggest_int("m6/repeat", 1, 5)
    m6_out_channel = trial.suggest_int("m6/out_channels", low=16, high=128, step=16)
    m6_stride = trial.suggest_int("m6/stride", low=1, high=UPPER_STRIDE)
    # force stride m6
    if n_stride == 2:
        m4_stride = 2
    if m6 == "Conv":
        # Conv args: [out_channel, kernel_size, stride, padding, groups, activation]
        m6_out_channel = trial.suggest_int("m6/out_channels", low=16, high=512, step=16)
        m6_kernel = trial.suggest_int("m6/kernel_size", low=1, high=5, step=2)
        m6_activation = trial.suggest_categorical("m6/activation", ["ReLU", "Hardswish"])
        m6_args = [m6_out_channel, m6_kernel, m6_stride, None, 1, m6_activation]
    elif m6 == "DWConv":
        # DWConv args: [out_channel, kernel_size, stride, padding_size, activation]
        m6_out_channel = trial.suggest_int("m6/out_channels", low=16, high=512, step=16)
        m6_kernel = trial.suggest_int("m6/kernel_size", low=1, high=5, step=2)
        m6_activation = trial.suggest_categorical("m6/activation", ["ReLU", "Hardswish"])
        m6_args = [m6_out_channel, m6_kernel, m6_stride, None, m6_activation]
    elif m6 == "MBConv":
        m6_c = trial.suggest_int("m6/c", low=16, high=320, step=16)
        m6_k = trial.suggest_int("m6/k", low=3, high=5, step=2)
        m6_args = [6, m6_c, m6_stride, m6_k]
    elif m6 == "InvertedResidualv2":
        m6_c = trial.suggest_int("m6/v2_c", low=16, high=128, step=16)
        m6_t = trial.suggest_int("m6/v2_t", low=1, high=8)
        m6_args = [m6_c, m6_t, m6_stride]
    elif m6 == "InvertedResidualv3":
        m6_kernel = trial.suggest_int("m6/kernel_size", low=3, high=5, step=2)
        m6_t = round(trial.suggest_float("m6/v3_t", low=1.0, high=6.0, step=0.1), 1)
        m6_c = trial.suggest_int("m6/v3_c", low=16, high=160, step=16)
        m6_se = trial.suggest_categorical("m6/v3_se", [0, 1])
        m6_hs = trial.suggest_categorical("m6/v3_hs", [0, 1])
        m6_args = [m6_kernel, m6_t, m6_c, m6_se, m6_hs, m6_stride]
    if not m6 == "Pass":
        if m6_stride == 2:
            n_stride += 1
            if n_stride>=MAX_NUM_STRIDE:
                UPPER_STRIDE = 1
        model.append([m6_repeat, m6, m6_args])

    # Module 7
    m7 = trial.suggest_categorical(
        "m7",
        ["Conv",
        "DWConv",
        "MBConv",
        "InvertedResidualv2",
        "InvertedResidualv3",
        "Pass"]
        )
    m7_args = []
    m7_repeat = trial.suggest_int("m7/repeat", 1, 5)
    m7_out_channel = trial.suggest_int("m7/out_channels", low=16, high=128, step=16)
    m7_stride = trial.suggest_int("m7/stride", low=1, high=UPPER_STRIDE)
    if m7 == "Conv":
        # Conv args: [out_channel, kernel_size, stride, padding, groups, activation]
        m7_out_channel = trial.suggest_int("m7/out_channels", low=128, high=1024, step=128)
        m7_kernel = trial.suggest_int("m7/kernel_size", low=1, high=5, step=2)
        m7_activation = trial.suggest_categorical("m7/activation", ["ReLU", "Hardswish"])
        m7_args = [m7_out_channel, m7_kernel, m7_stride, None, 1, m7_activation]
    elif m7 == "DWConv":
        # DWConv args: [out_channel, kernel_size, stride, padding_size, activation]
        m7_out_channel = trial.suggest_int("m7/out_channels", low=128, high=1024, step=128)
        m7_kernel = trial.suggest_int("m7/kernel_size", low=1, high=5, step=2)
        m7_activation = trial.suggest_categorical("m7/activation", ["ReLU", "Hardswish"])
        m7_args = [m7_out_channel, m7_kernel, m7_stride, None, m7_activation]
    elif m7 == "MBConv":
        m7_c = trial.suggest_int("m7/c", low=16, high=320, step=16)
        m7_k = trial.suggest_int("m7/k", low=3, high=5, step=2)
        m7_args = [6, m7_c, m7_stride, m7_k]
    elif m7 == "InvertedResidualv2":
        m7_c = trial.suggest_int("m7/v2_c", low=16, high=160, step=16)
        m7_t = trial.suggest_int("m7/v2_t", low=1, high=8)
        m7_args = [m7_c, m7_t, m7_stride]
    elif m7 == "InvertedResidualv3":
        m7_kernel = trial.suggest_int("m7/kernel_size", low=3, high=5, step=2)
        m7_t = round(trial.suggest_float("m7/v3_t", low=1.0, high=6.0, step=0.1), 1)
        m7_c = trial.suggest_int("m7/v3_c", low=8, high=160, step=8)
        m7_se = trial.suggest_categorical("m7/v3_se", [0, 1])
        m7_hs = trial.suggest_categorical("m7/v3_hs", [0, 1])
        m7_args = [m7_kernel, m7_t, m7_c, m7_se, m7_hs, m7_stride]
    if not m7 == "Pass":
        if m7_stride == 2:
            n_stride += 1
            if n_stride>=MAX_NUM_STRIDE:
                UPPER_STRIDE = 1
        model.append([m7_repeat, m7, m7_args])
        
    # last layer
    last_dim = trial.suggest_int("last_dim", low=128, high=512, step=128)
    # We can setup fixed structure as well
    model.append([1, "Conv", [last_dim, 1, 1]])
    model.append([1, "GlobalAvgPool", []])
    model.append([1, "FixedConv", [last_dim, 1, 1, None, 1, None]])
    return model


def tuning_score(test_f1: float, macs: float) -> float:
    f1_pivot = 0.85
    f1_limit = 0.5
    macs_pivot = 100000
    if test_f1 < f1_limit:
        score_f1 = 1
    elif f1_limit <= test_f1 < f1_pivot:
        score_f1 = 1 - (test_f1 / f1_pivot)
    else:
        score_f1 = 0.5 * (1 - (test_f1 / f1_pivot))
    score_macs = macs / macs_pivot
    result = score_f1 + score_macs
    return result


def objective(trial: optuna.trial.Trial, device) -> float:
    """Optuna objective.

    Args:
        trial
    Returns:
        float: tuning_score(accuracy & params)
    """
    model_config = copy.deepcopy(MODEL_CONFIG)
    data_config = copy.deepcopy(DATA_CONFIG)

    # hyperparams: EPOCHS, IMG_SIZE, n_select, BATCH_SIZE
    hyperparams = search_hyperparam(trial)

    model_config["input_size"] = [data_config["IMG_SIZE"], data_config["IMG_SIZE"]]
    # model_config["backbone"] = search_model(trial)

    # data_config["AUG_TRAIN_PARAMS"]["n_select"] = hyperparams["n_select"]
    data_config["BATCH_SIZE"] = hyperparams["BATCH_SIZE"]
    # data_config["EPOCHS"] = hyperparams["EPOCHS"]
    data_config["IMG_SIZE"] = hyperparams["IMG_SIZE"]
    log_dir = os.path.join("exp", datetime.now().strftime("%Y-%m-%d_%H-%M-%S"))
    os.makedirs(log_dir, exist_ok=True)
    model_instance = Model(model_config, verbose=False)
    macs = calc_macs(model_instance.model, (3, data_config["IMG_SIZE"], data_config["IMG_SIZE"]))

    # model_config, data_config
    _, test_f1, _ = train(
        model_config=model_config,
        data_config=data_config,
        log_dir=log_dir,
        fp16=data_config["FP16"],
        device=device,
    )
    return tuning_score(test_f1, macs)


def tune(gpu_id: int, storage: Union[str, None] = None, study_name: str = "pstage_automl"):
    if not torch.cuda.is_available():
        device = torch.device("cpu")
    elif 0 <= gpu_id < torch.cuda.device_count():
        device = torch.device(f"cuda:{gpu_id}")
    sampler = optuna.samplers.TPESampler(n_startup_trials=20)
    if storage is not None:
        rdb_storage = optuna.storages.RDBStorage(url=storage)
    else:
        rdb_storage = None
    study = optuna.create_study(
        directions=["minimize"],
        storage=rdb_storage,
        study_name=study_name,
        sampler=sampler,
        load_if_exists=True
    )
    study.optimize(lambda trial: objective(trial, device), n_trials=20)
    pruned_trials = [
        t for t in study.trials if t.state == optuna.trial.TrialState.PRUNED
    ]
    complete_trials = [
        t for t in study.trials if t.state == optuna.trial.TrialState.COMPLETE
    ]

    print("Study statistics: ")
    print("  Number of finished trials: ", len(study.trials))
    print("  Number of pruned trials: ", len(pruned_trials))
    print("  Number of complete trials: ", len(complete_trials))
    print("Best trials:")

    best_trials = study.best_trials
    # trials that satisfies Pareto Fronts
    for tr in best_trials:
        print(f"  value1:{tr.values[0]}, value2:{tr.values[1]}")
        for key, value in tr.params.items():
            print(f"    {key}:{value}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Optuna tuner.")
    parser.add_argument(
        "--gpu", default=0, type=int, help="GPU id to use"
    )
    parser.add_argument(
        "--storage", default="", type=str, help="RDB Storage URL for optuna."
    )
    parser.add_argument(
        "--study-name", default="pstage_automl", type=str, help="Optuna study name."
    )
    args = parser.parse_args()
    
    tune(args.gpu, storage=None if args.storage == "" else args.storage, study_name=args.study_name)
